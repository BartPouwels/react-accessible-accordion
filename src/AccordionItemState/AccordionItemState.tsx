import * as React from 'react';

type AccordionItemStateProps = React.HTMLAttributes<HTMLDivElement> & {
    expanded: boolean;
    children(expanded: boolean): React.ReactNode;
};

const AccordionItemState = (props: AccordionItemStateProps): JSX.Element => {
    return <>{props.children(props.expanded)}</>;
};
export default AccordionItemState;
