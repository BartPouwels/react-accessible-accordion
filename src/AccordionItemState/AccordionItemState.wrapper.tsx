import * as React from 'react';
import * as propTypes from '../helpers/propTypes';

import {
    CONTEXT_KEY as ACCORDION_CONTEXT_KEY,
    getAccordionStore,
    Item,
} from '../AccordionContainer/AccordionContainer';
import {
    CONTEXT_KEY as ITEM_CONTEXT_KEY,
    getItemStore,
} from '../ItemContainer/ItemContainer';

import AccordionItemState from './AccordionItemState';

type AccordionItemStateWrapperProps = React.HTMLAttributes<HTMLDivElement> & {
    expanded?: boolean;
    render(expanded: boolean): React.ReactNode;
};

type AccordionItemStateWrapperContext = {
    [ACCORDION_CONTEXT_KEY](): null;
    [ITEM_CONTEXT_KEY](): null;
};

export default class AccordionItemStateWrapper extends React.Component<
    AccordionItemStateWrapperProps,
    AccordionItemStateWrapperContext
> {
    static contextTypes: AccordionItemStateWrapperContext = {
        [ACCORDION_CONTEXT_KEY]: propTypes.wildcard,
        [ITEM_CONTEXT_KEY]: propTypes.wildcard,
    };

    static defaultProps: AccordionItemStateWrapperProps = {
        render: (expanded: boolean): React.ReactNode => null,
    };

    render(): JSX.Element {
        const accordionStore = getAccordionStore(this.context);

        if (!accordionStore) {
            // tslint:disable-next-line:no-console
            console.error(
                'AccordionItemState component cannot render because it has not been nested inside an Accordion component.',
            );

            return null;
        }

        const itemStore = getItemStore(this.context);

        if (!itemStore) {
            // tslint:disable-next-line:no-console
            console.error(
                'AccordionItemState component cannot render because it has not been nested inside an AccordionItem component.',
            );

            return null;
        }

        const { uuid } = itemStore;
        const { items } = accordionStore;
        const item = items.filter(
            (stateItem: Item) => stateItem.uuid === uuid,
        )[0];
        const { render } = this.props;

        return item ? (
            <AccordionItemState expanded={item.expanded} render={render} />
        ) : null;
    }
}
