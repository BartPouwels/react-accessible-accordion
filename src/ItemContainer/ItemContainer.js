// @flow
import { Container } from 'unstated';
import consecutive from 'consecutive';

export type StoreState = {
    uuid: string | number,
};

let nextUuid = consecutive();
export function resetNextUuid() {
    nextUuid = consecutive();
}

class ItemContainer extends Container<StoreState> {
    state = {
        uuid: nextUuid(),
    };
}

export default ItemContainer;
