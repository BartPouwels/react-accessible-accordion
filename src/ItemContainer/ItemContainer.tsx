// tslint:disable:max-classes-per-file

import * as React from 'react';

export type UUID = string | number;

export type ProviderProps = {
    children?: React.ReactNode;
    uuid: UUID;
};

export type ItemContainer = {
    uuid: UUID;
};

// Arbitrary, but ought to be unique to avoid context namespace clashes.
export const CONTEXT_KEY = 'react-accessible-accordion@ItemContainer';

export class Provider extends React.Component<ProviderProps> {
    static childContextTypes = {
        // Empty anonymous callback is a hacky 'wildcard' workaround for bypassing prop-types.
        [CONTEXT_KEY]: () => null,
    };

    getChildContext() {
        const { uuid } = this.props;
        const context: ItemContainer = {
            uuid,
        };

        return {
            [CONTEXT_KEY]: context,
        };
    }

    render() {
        return this.props.children || null;
    }
}

type ConsumerProps = {
    children(container: ItemContainer): React.ReactNode;
};

export class Consumer extends React.Component<ConsumerProps> {
    static contextTypes = {
        // Empty anonymous callback is a hacky 'wildcard' workaround for bypassing prop-types.
        [CONTEXT_KEY]: () => null,
    };

    render() {
        return this.props.children(this.context[CONTEXT_KEY]);
    }
}

export const getItemStore = (context): ItemContainer | undefined => {
    return context[CONTEXT_KEY];
};
