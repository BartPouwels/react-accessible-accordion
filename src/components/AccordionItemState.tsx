import * as React from 'react';
import { DivAttributes } from '../helpers/types';
import { Consumer as ItemConsumer, ItemContext } from './ItemContext';

type Props = Pick<DivAttributes, Exclude<keyof DivAttributes, 'children'>> & {
    children(expanded?: boolean): React.ReactNode;
};

export default class AccordionItemState extends React.Component<Props> {
    renderChildren = (itemContext: ItemContext): JSX.Element => {
        const { expanded } = itemContext;

        return <React.Fragment>{this.props.children(expanded)}</React.Fragment>;
    };

    render(): JSX.Element {
        return <ItemConsumer>{this.renderChildren}</ItemConsumer>;
    }
}
