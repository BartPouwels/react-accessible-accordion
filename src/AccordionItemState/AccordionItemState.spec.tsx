import { mount, ReactWrapper } from 'enzyme';
import * as React from 'react';
import {
    Item,
    Provider as AccordionProvider,
} from '../AccordionContainer/AccordionContainer';
import { Provider as ItemProvider } from '../ItemContainer/ItemContainer';
import { default as AccordionItemState } from './AccordionItemState.wrapper';

describe('AccordionItemState', () => {
    function mountItem(children: React.ReactNode): ReactWrapper {
        const item: Item = {
            uuid: 0,
            expanded: true,
        };

        return mount(
            <AccordionProvider items={[item]}>
                <ItemProvider uuid={item.uuid}>{children}</ItemProvider>
            </AccordionProvider>,
        );
    }

    it('renders correctly with min params', () => {
        const wrapper = mountItem(<AccordionItemState />);
        expect(wrapper).toMatchSnapshot();
    });

    it('does not render if no accordionStore found in context', () => {
        const wrapper = mount(
            <ItemProvider uuid="foo">
                <AccordionItemState />
            </ItemProvider>,
        );

        expect(wrapper.find('div[data-enzyme]').length).toEqual(0);
    });

    it('does not render if no itemStore found in context', () => {
        const wrapper = mount(
            <AccordionProvider items={[]}>
                <AccordionItemState />
            </AccordionProvider>,
        );

        expect(wrapper.find('div[data-enzyme]').length).toEqual(0);
    });

    it('renders correctly with different children prop', () => {
        const renderProp = (expanded: boolean): React.ReactNode =>
            expanded ? (
                <div className="expanded" />
            ) : (
                <div className="collapsed" />
            );

        const wrapper = mountItem(<AccordionItemState children={renderProp} />);

        expect(wrapper.find('div').hasClass('expanded')).toEqual(true);
        expect(wrapper.find('div').hasClass('collapsed')).toEqual(false);
    });

    it('renders correctly with different children prop and expanded set to false', () => {
        function mountExpandedFalse(children: React.ReactNode): ReactWrapper {
            const item: Item = {
                uuid: 0,
                expanded: false,
            };

            return mount(
                <AccordionProvider items={[item]}>
                    <ItemProvider uuid={item.uuid}>{children}</ItemProvider>
                </AccordionProvider>,
            );
        }

        const renderProp = (expanded: boolean): React.ReactNode =>
            expanded ? (
                <div className="expanded" />
            ) : (
                <div className="collapsed" />
            );

        const wrapper = mountExpandedFalse(
            <AccordionItemState children={renderProp} />,
        );

        expect(wrapper.find('div').hasClass('expanded')).toEqual(false);
        expect(wrapper.find('div').hasClass('collapsed')).toEqual(true);
    });
});
