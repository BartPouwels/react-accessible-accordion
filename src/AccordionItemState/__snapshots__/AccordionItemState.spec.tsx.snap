// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItemState renders correctly with min params 1`] = `""`;
